nonthermal.cc

```diff
@@ -2682,23 +2739,17 @@ void nt_MPI_Bcast(const int modelgridindex, const int root) {
     MPI_Bcast(nt_solution[modelgridindex].ionenfrac_num_auger, get_includedions() * (NT_MAX_AUGER_ELECTRONS + 1),
               MPI_FLOAT, root, MPI_COMM_WORLD);
 
-    // communicate NT excitations
-    const auto frac_excitations_list_size_old = nt_solution[modelgridindex].frac_excitations_list.size();
-    auto frac_excitations_list_size_new = nt_solution[modelgridindex].frac_excitations_list.size();
+    // communicate NT excitation size to all processes
+    auto frac_excitations_list_size_new = nt_solution[modelgridindex].frac_excitations_list_size;
     MPI_Bcast(&frac_excitations_list_size_new, 1, MPI_INT, root, MPI_COMM_WORLD);
 
-    if (frac_excitations_list_size_new != frac_excitations_list_size_old) {
-      nt_solution[modelgridindex].frac_excitations_list.resize(frac_excitations_list_size_new);
-    }
+    nt_solution[modelgridindex].frac_excitations_list_size = frac_excitations_list_size_new;
 
-    const auto frac_excitations_list_size = nt_solution[modelgridindex].frac_excitations_list.size();
-    for (size_t excitationindex = 0; excitationindex < frac_excitations_list_size; excitationindex++) {
-      MPI_Bcast(&nt_solution[modelgridindex].frac_excitations_list[excitationindex].frac_deposition, 1, MPI_DOUBLE,
-                root, MPI_COMM_WORLD);
-      MPI_Bcast(&nt_solution[modelgridindex].frac_excitations_list[excitationindex].ratecoeffperdeposition, 1,
-                MPI_DOUBLE, root, MPI_COMM_WORLD);
-      MPI_Bcast(&nt_solution[modelgridindex].frac_excitations_list[excitationindex].lineindex, 1, MPI_INT, root,
-                MPI_COMM_WORLD);
+    if (globals::rank_in_node == 0) {
+      // communicate NT excitation list via inter-node communication
+      MPI_Bcast(&nt_solution[modelgridindex].frac_excitations_list,
+                nt_solution[modelgridindex].frac_excitations_list_size * sizeof(nt_excitation_struct), MPI_BYTE,
+                root_node_id, globals::mpi_comm_internode);
     }
```

Comment:
There is something wrong with this bcast. I have run a 1D model, and on timestep 8, on the first time going through this function, the code crashes. Is there something simple thats wrong with this bcast that I'm missing? There are sensible values inside the frac_excitations_list, so its not that the values are infs etc

Reply:
It looks like there’s an extra ampersand on the first argument.

Reply:
I’m surprised that this didn’t crash the CI job…it’s worth figuring that out because we probably need a better test.

Reply:
That fixed the problem. I had assumed we needed the ampersand to point to a specific memory address, but apparently not.

Reply:
frac_excitations_list is a already pointer type, so taking the address of it gives you a pointer to the pointer.